### ISSUE TITLE
Design Q:  cache2k on top of LMDB

### ISSUE THREAD

**Issue by vladp** [cache2k/cache2k#98]
Hello.
I am using cache2k to cache  responses from database, while running within Jetty container. I am still far from production. But this  works quite well and allows me the granularity of controls I need.

I am also using LMDB (KV value store) via https://github.com/lmdbjava/lmdbjava   to 'near-cache' files.
LMDB, being a KV database offers all the necessary thread-safe transaction and access semantic for KV, but it does not have built in features that are cache specific (eg expiration/etc).

Would be possible/reasonable for me try to integrate cache2k with lmdb-java?

Is there a good starting project/effort for me look in to?

**Comment by cruftex** [comment]
There are different variants to integrate the cache and the database.

1. Use the cache inside the DB code to reduce latency. Which means that hot data is kept in the heap instead of reading it all the time from disk. Usually a database has such a kind of cache, but I did not investigate LMDB yet. Using cache2k in that context can make sense since it is highly efficient and typically faster then any other cache implementation. The cache will be internal to LMDB.

2. Use LMDB has a persistence storage for cache2k. The current cache2k version has no persistence feature or no interface to allow to build caching hierarchies with all features. There is some earlier work on that. That is tracked here: #30.

3. It's possible to build simple hierarchies with the `CacheLoader` and `CacheWriter` interface. We use that to integrate with DBs, but it is always targeted for a specific use case.

**Comment by vladp** [comment]
thank you.

It is the variant 2 that I was thinking about. As it would fit the use case best (letting LMDB to cache files, while using cache2K semantic for expiration/cache size maintenance and cache-stats reporting).

Looking at the challenges you described at #30, it seems like this would be a bit too challenging for me to solve.  But it was nice to read that you were thinking/contemplating the extensibility of cache2k in that way.

a bit more background on LMDB:


LMDB is one of the worlds fastest embedded (non-distributed/single host)  KV stores that leverages OS memory mapped files. While maintaining thread-safety and transactional consistency of updates. 

Basically to conserve ram-memory, and to avoid file system looks ups through deep and 'wide' directory structures (including remote disks), I am using LMDB to leverage OS virtual memory/paging to cache specific files for faster serving.

Open source tools such as[ Intel's Caffe ](https://github.com/BVLC/caffe) (deep learning framework for image recognition), uses LMDB (optionally) to store image file files for training sets (because in those example, the same image must be processed/evaluated many many times, and it also simplifies storing several million small images,  avoiding expensive directory lookups/extracting files onto disk).
